### DISCUSSION TITLE
mcservice does not start on a fresh Ubuntu 20.04 install

### DISCUSSION THREAD

**Discussion by darkstarfish1337** [nextgenhealthcare/connect#4617] [Q&A]
Hello,

trying to get this mirth server to run. Installed it on Ubuntu 20.04 with openjdk-16-jdk. Sysctemctl status mcservice shows me the following:

● mcservice.service - Mirth Connect Service
     Loaded: loaded (/etc/systemd/system/mcservice.service; enabled; vendor preset: enabled)
     Active: failed (Result: exit-code) since Fri 2021-06-25 08:11:08 UTC; 3min 22s ago
    Process: 554 ExecStart=/opt/mirthconnect/mcservice start-launchd (code=exited, status=1/FAILURE)
   Main PID: 554 (code=exited, status=1/FAILURE)

Jun 25 08:11:08 srv254 mcservice[554]: Caused by: java.lang.reflect.InaccessibleObjectException: Unable to make field private final java.util.Comparator java.util.TreeMap.comparator accessible: module java.base does not "opens java.util" to unnamed module @17bffc17
Jun 25 08:11:08 srv254 mcservice[554]:         at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:357)
Jun 25 08:11:08 srv254 mcservice[554]:         at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:297)
Jun 25 08:11:08 srv254 mcservice[554]:         at java.base/java.lang.reflect.Field.checkCanSetAccessible(Field.java:177)
Jun 25 08:11:08 srv254 mcservice[554]:         at java.base/java.lang.reflect.Field.setAccessible(Field.java:171)
Jun 25 08:11:08 srv254 mcservice[554]:         at com.thoughtworks.xstream.core.util.Fields.locate(Fields.java:40)
Jun 25 08:11:08 srv254 mcservice[554]:         at com.thoughtworks.xstream.converters.collections.TreeMapConverter.<clinit>(TreeMapConverter.java:50)
Jun 25 08:11:08 srv254 mcservice[554]:         ... 31 more
Jun 25 08:11:08 srv254 systemd[1]: mcservice.service: Main process exited, code=exited, status=1/FAILURE
Jun 25 08:11:08 srv254 systemd[1]: mcservice.service: Failed with result 'exit-code'.

Am I using the wrong Java environment? What's going on here?

Kind regards,
Starfish

**Comment by pacmano1** [comment]
Installed how?  With the installer?   The installer runs mirth as root.   That's bad practice.  I'd use the tar.gz file and create a named user that owns the extracted file system.   

See https://forums.mirthproject.io/forum/mirth-connect/support/178428-run-mirth-service-manually-works-running-from-systemd-errs-with-no-suitable-jvm?p=179090#post179090 for a systemd start up script.

**Comment by darkstarfish1337** [comment]
I tried to follow your hints. 

- Reinstalled Ubuntu 20.04
- Installed openjdk-16-jdk

java -version
openjdk version "16.0.1" 2021-04-20
OpenJDK Runtime Environment (build 16.0.1+9-Ubuntu-120.04)
OpenJDK 64-Bit Server VM (build 16.0.1+9-Ubuntu-120.04, mixed mode, sharing)

- Created a new linux user named "mirth"
- Downloaded and extracted the tar into /opt/MC/  which is owned by mirth with wrx rights
- Created the following mirth.service systemd file 

[Unit]
Description=mirth
After=syslog.target

[Service]
Type=simple
RemainAfterExit=yes
Environment=PATH=/usr/lib/jvm/java-16-openjdk-amd64/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
ExecStart=/opt/MC/mcserver

[Install]
WantedBy=multi-user.target

- systemctl daemon-reload
- systemd-analyze mirth.service -> no errors
- systemctl enable mirth.service -> no errors
- systemctl start mirth.service -> no errors

I also tried to understand the INSTALL4J_JAVA_HOME bit from the thread, but that's a different issue.  However, I get the same error as before when I check the status of the mirth.service. 

● mirth.service - mirth
     Loaded: loaded (/lib/systemd/system/mirth.service; enabled; vendor preset: enabled)
     Active: failed (Result: exit-code) since Thu 2021-07-01 13:34:21 UTC; 2s ago
    Process: 214664 ExecStart=/opt/MC/mcserver (code=exited, status=1/FAILURE)
   Main PID: 214664 (code=exited, status=1/FAILURE)

Jul 01 13:34:21 srv254 mcserver[214664]: Caused by: java.lang.reflect.InaccessibleObjectException: Unable to make field private final java.util.Comparator java.util.TreeMap.comparator accessible: >
Jul 01 13:34:21 srv254 mcserver[214664]:         at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:357)
Jul 01 13:34:21 srv254 mcserver[214664]:         at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:297)
Jul 01 13:34:21 srv254 mcserver[214664]:         at java.base/java.lang.reflect.Field.checkCanSetAccessible(Field.java:177)
Jul 01 13:34:21 srv254 mcserver[214664]:         at java.base/java.lang.reflect.Field.setAccessible(Field.java:171)
Jul 01 13:34:21 srv254 mcserver[214664]:         at com.thoughtworks.xstream.core.util.Fields.locate(Fields.java:40)
Jul 01 13:34:21 srv254 mcserver[214664]:         at com.thoughtworks.xstream.converters.collections.TreeMapConverter.<clinit>(TreeMapConverter.java:50)
Jul 01 13:34:21 srv254 mcserver[214664]:         ... 31 more
Jul 01 13:34:21 srv254 systemd[1]: mirth.service: Main process exited, code=exited, status=1/FAILURE
Jul 01 13:34:21 srv254 systemd[1]: mirth.service: Failed with result 'exit-code'.

**Reply by pacmano1** [reply]
I have not used v16 jdk.

**Reply by darkstarfish1337** [reply]
Which version are you using? 

openjdk v14 has the following warnings

● mirth.service - mirth
     Loaded: loaded (/lib/systemd/system/mirth.service; enabled; vendor preset: enabled)
     Active: active (running) since Thu 2021-07-01 14:14:41 UTC; 18s ago
   Main PID: 230391 (java)
      Tasks: 40 (limit: 2279)
     Memory: 324.2M
     CGroup: /system.slice/mirth.service
             └─230391 /usr/lib/jvm/java-14-openjdk-amd64/bin/java -server -Xmx256m -Djava.awt.headless=true -Dapple.awt.UIElement=true -classpath /opt/MC/.install4j/i4jruntime.jar:/opt/MC/.install>

Jul 01 14:14:41 srv254 systemd[1]: Started mirth.
Jul 01 14:14:42 srv254 mcserver[230391]: WARNING: An illegal reflective access operation has occurred
Jul 01 14:14:42 srv254 mcserver[230391]: WARNING: Illegal reflective access by com.thoughtworks.xstream.core.util.Fields (file:/opt/MC/server-lib/xstream-1.4.12.jar) to field java.util.TreeMap.com>
Jul 01 14:14:42 srv254 mcserver[230391]: WARNING: Please consider reporting this to the maintainers of com.thoughtworks.xstream.core.util.Fields
Jul 01 14:14:42 srv254 mcserver[230391]: WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
Jul 01 14:14:42 srv254 mcserver[230391]: WARNING: All illegal access operations will be denied in a future release

**Reply by darkstarfish1337** [reply]
Seems to be running with v14. Java however is running via root. Wouldn't we want to have the underlying java processes not running as root as well?

**Reply by pacmano1** [reply]
Can't you add User= and Group= in the service section?

**Reply by darkstarfish1337** [reply]
> 
> 
> Can't you add User= and Group= in the service section?

I tried this and got all the processes to run under a normal user.  When checking the status of mirth.service I get the following message:

Jul 12 14:21:53 srv254 systemd[1]: Started mirth.
Jul 12 14:21:55 srv254 mcserver[38796]: WARNING: An illegal reflective access operation has occurred
Jul 12 14:21:55 srv254 mcserver[38796]: WARNING: Illegal reflective access by com.thoughtworks.xstream.core.util.Fields (file:/opt/mirth/server-lib/xstream-1.4.12.jar) to field java.util.TreeMap.comparator
Jul 12 14:21:55 srv254 mcserver[38796]: WARNING: Please consider reporting this to the maintainers of com.thoughtworks.xstream.core.util.Fields
Jul 12 14:21:55 srv254 mcserver[38796]: WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
Jul 12 14:21:55 srv254 mcserver[38796]: WARNING: All illegal access operations will be denied in a future release

Is this warning of "Illegal reflective access by com.thoughtworks.xstream.core.util.Fields (file:/opt/mirth/server-lib/xstream-1.4.12.jar) to field java.util.TreeMap.comparator" somthing to warn about? Am I overlooking something here?

**Reply by jonbartels** [reply]
#4624 is related to the warnings for XStream. #4602 is related for any fatal issues starting up under Java 16.
